tsdb/querier.go

```diff
@@ -778,18 +778,12 @@ func (m *mergedStringIter) Next() bool {
 		return false
 	}
 
-	if !m.aok {
+	if !m.aok || m.a.At() > m.b.At() {
```

Comment:
This does not seem safe since at this stage g might not be ok

Reply:
FWIW this is an issue in the existing code, right? (should be fixed either way, just noting)

When `!m.aok` it's gonna drop into this loop, which does the same `m.b.At()` (as well as `m.b.Next()`), without any check on `m.bok`.

Reply:
It is not an issue with the current code.

This code first check that aok and bok and true.

This change removes that check and go directly with the assumption they are true.

Reply:
Hmm, am I really misinterpreting this?

```go
if !m.aok {
		m.cur = m.b.At()
		m.bok = m.b.Next()
else if !m.bok {
}
```

The first check only checks `aok` - if it is not ok, it then operates on `b` (`m.cur = m.b.At()`) without checking for `bok`. So if `aok` is false and `bok` is false, it'll still drop in here (I'm assuming here that it is possible for both to be !ok simultaneously, considering your comment above - `since at this stage b might not be ok`)

Reply:
if !bok we would never check if m.a.At() > m.b.At()

Reply:
Oh right, that's my miscomprehension - so the problem arises in the comparison, not the call on `m.b`. Cool, excuse my derailing there - I thought it was unsafe to act on `m.b`.

Reply:
I am not an expert in this code, so I just wanted to make clear that the code is not equivalent from a pure go perspective.

Reply:
I think the current logic is fine and there is no issue here. Accessing `m.b` without checking for `m.bok` seems wrong. Assume the `m.b` was exhausted with `m.bok` being false, and if `m.b.At()` always gives the last value, then `m.a.At() > m.b.At()` is always going to be true hence go into infinite loop of looking at the last value in `m.b`. Same thing when `m.a` is exhausted for the next condition.